File: 0022-generate-parentheses/0022-generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    let result = [];
    backtrack( "", 0, 0 ,n);
    return result;

    function backtrack( curr, open , close, m){
        if(curr.length === 2 * m){
            result.push(curr);
        }
        if(open < n){
            backtrack( curr + "(", open + 1, close, m);
        }
        if(close < open){
            backtrack( curr + ")", open, close + 1 ,m);
        }
    }
    
};

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        backtrack(result, "" ,0, 0, n);
        return result;   
    }

    public void backtrack(List<String> result, String current, int open, int close, int n ){
        if(current.length() == 2*n){
            result.add(current);
        }
        if(open < n){
            backtrack(result, current + "(", open + 1, close, n);
        }
        if(close < open){
            backtrack(result, current + ")", open, close + 1, n);
        }
    }
}